src/Workspaces/Core/Portable/Workspace/Solution/SourceGeneratedDocumentIdentity.cs

```diff
@@ -0,0 +1,64 @@
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// A small struct that holds the values that define the identity of a source generated document, and don't change
+    /// as new generations happen. This is mostly for convenience as we are reguarly working with this combination of values.
+    /// </summary>
+    internal readonly struct SourceGeneratedDocumentIdentity
+    {
+        public SourceGeneratedDocumentIdentity(DocumentId documentId, string hintName, ISourceGenerator generator, string filePath)
+        {
+            DocumentId = documentId;
+            HintName = hintName;
+            Generator = generator;
+            FilePath = filePath;
+        }
+
+        public DocumentId DocumentId { get; }
+        public string HintName { get; }
+        public ISourceGenerator Generator { get; }
+        public string FilePath { get; }
+
+        public static SourceGeneratedDocumentIdentity Generate(ProjectId projectId, string hintName, ISourceGenerator generator, string filePath)
+        {
+            // We want the DocumentId generated for a generated output to be stable between Compilations; this is so features that track
+            // a document by DocumentId can find it after some change has happened that requires generators to run again.
+            // To achieve this we'll just do a crytographic hash of the generator name and hint name; the choice of a cryptographic hash
+            // as opposed to a more generic string hash is we actually want to ensure we don't have collisions.
+            var generatorName = generator.GetType().FullName;
+            Contract.ThrowIfNull(generatorName);
+
+            // Combine the strings together; we'll use Encoding.Unicode since that'll match the underlying format; this can be made much
+            // faster once we're on .NET Core since we could directly treat the strings as ReadOnlySpan<char>.
+            var projectIdBytes = projectId.Id.ToByteArray();
+            using var _ = ArrayBuilder<byte>.GetInstance(capacity: (generatorName.Length + hintName.Length + 1) * 2 + projectIdBytes.Length, out var hashInput);
+            hashInput.AddRange(projectIdBytes);
+            hashInput.AddRange(Encoding.Unicode.GetBytes(generatorName));
+
+            // Add a null to separate the generator name and hint name; since this is effectively a joining of UTF-16 bytes
+            // we'll use a UTF-16 null just to make sure there's absolutely no risk of collision.
+            hashInput.AddRange(0, 0);
+            hashInput.AddRange(Encoding.Unicode.GetBytes(hintName));
+
+            // The particular choice of crypto algorithm here is arbitrary and can be always changed as necessary. The only requirement
+            // is it must be collision resistant, and provide enough bits to fill a GUID.
+            using var crytpoAlgorithm = System.Security.Cryptography.SHA256.Create();
+            var hash = crytpoAlgorithm.ComputeHash(hashInput.ToArray());
```

Comment:
use of the array builder seems unnecessary, esp. as you are toarray'ing it.  don't the crypto functions allow you to pass many values along cominbing to a final hash?

Reply:
So there is https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.hashalgorithm.transformblock?view=net-5.0 which seemed somewhat tricky to use since there's an output buffer happening in addition to the inputs. But maybe I'm misreading it or something. I'll take this as a follow-up item to look at further (this code is moving for now.)